hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java

```diff
@@ -361,15 +372,38 @@ private boolean isNewInstantBlock(HoodieLogBlock logBlock) {
    */
   private void processDataBlock(HoodieDataBlock dataBlock, Option<List<String>> keys) throws Exception {
     try (ClosableIterator<IndexedRecord> recordItr = dataBlock.getRecordItr(keys.orElse(Collections.emptyList()))) {
+      boolean headRecord = true;
+      Option<Schema> schemaOption = Option.empty();
       while (recordItr.hasNext()) {
-        IndexedRecord record = recordItr.next();
+        IndexedRecord currentRecord = recordItr.next();
+        if (headRecord) {
+          schemaOption = getMergedSchema(dataBlock, currentRecord);
+          headRecord = false;
+        }
+        IndexedRecord record = schemaOption.isPresent() ? AvroSchemaUtil.rewriteRecord(currentRecord, schemaOption.get()) : currentRecord;
         processNextRecord(createHoodieRecord(record, this.hoodieTableMetaClient.getTableConfig(), this.payloadClassFQN,
             this.preCombineField, this.withOperationField, this.simpleKeyGenFields, this.partitionName));
         totalLogRecords.incrementAndGet();
       }
     }
   }
 
+  private Option<Schema> getMergedSchema(HoodieDataBlock dataBlock, IndexedRecord record) {
```

Comment:
This is an important place where we can add descriptive java-docs around this method .

Reply:
added

Reply:
fixed